src/app/chat/chat.component.ts

```diff
@@ -27,34 +27,60 @@ export class ChatComponent {
     userId: string | null;
     username: string | null;
   }> = this.store.select(selectUserData);
-  protected readonly datasource: IDatasource<Message> = this.getDatasource();
+
+  protected readonly datasource = new Datasource<Message>({
+    get: (index: number, count: number) => {
+      console.log(index, count);
+      return this.store
+        .select(selectMessagesSlice(index, count))
+        .pipe(take(1), tap(console.log));
+    },
+    devSettings: { debug: true },
+  });
 
   constructor(
     private fb: FormBuilder,
     private store: Store<EntityState<Message>>
   ) {
+    let prev: Message[] = [];
     this.messages$
       .pipe(
-        tap(messages =>
-          this.datasource.adapter?.reset(this.getDatasource(messages.length))
-        )
+        tap(async curr => {
+          const newItem = curr.find(c => !prev.find(p => c.id == p.id));
+          if (newItem) {
+            prev = curr;
+            this.appendMessage(newItem);
+            return;
+          }
+          const oldItem = curr.find(c =>
+            prev.find(p => c.id == p.id && c.time !== p.time)
+          );
+          if (oldItem) {
+            prev = curr;
+            this.replaceMessage(oldItem);
+            return;
+          }
+        })
       )
       .subscribe();
   }
 
-  getDatasource<T>(maxIndex?: number): IDatasource<T> {
-    return new Datasource<T>({
-      get: (index: number, count: number) => {
-        // why is index always 1?
-        // why does index start at 1?
-        console.log(index, count);
-        return this.store
-          .select(selectMessagesSlice(index, count))
-          .pipe(take(1), tap(console.log));
-      },
-      settings: {
-        maxIndex: maxIndex ?? Infinity,
-      },
+  async appendMessage(message: Message) {
+    await this.datasource.adapter.relax();
+    await this.datasource.adapter.append({
+      items: [message],
+      eof: true,
+    });
+    await this.datasource.adapter.clip({
+      forwardOnly: true,
+    });
+  }
```

Comment:
So, we use Adapter.append to push new item to the end of the list on the UI. But if the end-of-file is not reached, this push will be virtual (`eof: true` provides such behaviour). Also, we are clipping items after append to provide virtualization over unlimited appending. Clip will break through the end-of-file at some point, and new items will be virtualized.